File: Scala/ćw/List3.scala
//mapy
List(1,2,3) map ((n:Int) => n*10)
List[Iny] = (10,20,30)

//filtry
List(2,4,6) filter ((n:Int) => n>3)
res4: List[Int] = List(4, 6)

//zad2
// Zdefiniuj funkcje a) curry3 i b) uncurry3, przeprowadzaj¹ce konwersjê miêdzy
//zwiniêtymi i rozwiniêtymi postaciami funkcji od trzech argumentów. Podaj ich typy. 
def uncurry3[a1, a2, a3, b](f: a1 => a2 => a3 => b): (a1, a2, a3) => b =
  (x1, x2, x3) => f(x1)(x2)(x3)
  
def curry3[a1, a2, a3, b](f: (a1, a2, a3) => b): a1 => a2 => a3 => b =
  x1 => x2 => x3 => f(x1, x2, x3)

//test
val add_rozwiniete  = (x:Int, y: Int, z:Int) => x+y+z
uncurry3 (add_rozwiniete) //zwijam
((add curried 1) (2)) (5) 
 
val add_zwiniete  = (x:Int) => (y:Int) => (z:Int) => x+y+z
curry3 (add_zwiniete) //rozwijam

//zad 3
//sumProd, która oblicza nierekurencyjnie jednoczeœnie sumê i iloczyn listy liczb 
//ca³kowitych z jednokrotnym u¿yciem funkcji bibliotecznej fold_left

val list = List(2,3,4)

//scala> def f(a: (Int,Int), b:Int) = (a._1+b,a._2*b)
//f: (a: (Int, Int), b: Int)(Int, Int)

//scala> list.foldLeft(0,1)(f)
//res8: (Int, Int) = (9,24)

 def sumProd(xs:List[Int]):(Int,Int) = 
   xs match{
      case (Nil) => (0,1)
      case _ => xs.foldLeft(0,1)((a: (Int,Int), b:Int) => (a._1+b,a._2*b))
  }
  
 def sumProd2(xs:List[Int]):(Int,Int) = {
   xs.foldLeft(0,1)((a: (Int,Int), b:Int) => (a._1+b,a._2*b))
 }
 
 def sumProd3(l:List[Int]) = 
	l.foldLeft(0,1) ((s:(Int,Int),h:Int) =>
    	s match {
        	case (s,p) => (h+s, h*p)
  })

// Zad 4
def isGreater(a: Int, b: Int) = a > b

def insertionSort[A](isGreater: (A, A) => Boolean, list: List[A]): List[A] = {
    def insert(element: A, list: List[A]): List[A] = {
      list match {
        case Nil    => List(element)
        case h :: t =>
          if (isGreater(element, h)) 
	                      h :: insert(element, t)
          else element :: h :: t
      }
    }
  list match {
    case Nil    => Nil
    case h :: t => insert(h, insertionSort(isGreater, t))
  }
}
insertionSort(isGreater, List(5,4,1,3,2,5,3,3))

//python (???)
def insertionSort(isGreater, elements):
  def insert(e, elements):
    if len(elements) ==0:
      return [e, ]
    elif isGreater(e, elements[0]):
      return elements[0:] +insert(e, elements[1:]
    else:
      return [e, ] + elements
  if len(elements) == 0:
     return []
  else:
     return insert(elements[0], elements[1:])
// b

File: OCaml/ćw/List3.ml
(*mapy*)
List.map (fun x -> x*x) [1;3;4;5]
-: int list = [1;9;16;25]

List.map String.length ["ala"; "y"; "toto"]
-: int list = [3;1;4]

(*filtry*)

List.filter (fun s -> String.length s >= 3)  ["ala"; "y"; "toto"]
-: string list = ["ala";"toto"]

(*zad1*)
a)
let f x = x 1 1;;
let f = fun x -> x 1 1;;

(*po pierwsze x nie wywoluje funkcjitylko jest funkcja
returnem jaki f otrzyma bd wynik x 1 1, a nie wiadomo co to jest wiemy tylko,
ze x jest w postaci rozwinietejx: 1->1->niewiadomo *)
(int -> int -> a') -> a'

b)
let f2 x y z = x ( y ^ z );;
let f2 = fun x -> fun y -> fun z -> x ( y^z );;

(* x jest funkcja ktora przyjmuje string
y i z tez sa stringami
patrze na x no a ona przyjmuje argument string bo masz konkatenacje
(String  a-> 'a)->string b->string c->'a *)
(string -> 'a) -> string -> string -> 'a


(*zad2*)
let curry3 f x y z = f (x, y, z);;
let curry = function f -> function x -> function y -> function z -> f (x, y, z)
(* val curry3 : ('a * 'b * 'c -> 'd) -> 'a -> 'b -> 'c -> 'd = <fun> *)

let uncurry3 f (x, y, z) = f x y z;;
(* val uncurry3 : ('a -> 'b -> 'c -> 'd) -> 'a * 'b * 'c -> 'd = <fun> *)


(*zad3*)
(*sumProd, która oblicza nierekurencyjnie jednocześnie sumę i iloczyn listy liczb 
całkowitych z jednokrotnym użyciem funkcji bibliotecznej fold_left *)
let sumProd xs =
		match xs with
		 [] -> (0,1)
		| _ -> List.fold_left (fun a b -> fst a+b, snd a*b) (0, 1) xs

let sumProd2 xs =
	 List.fold_left (fun a, b -> fst a+b, snd a*b) (0, 1) xs
	
let sumProd2 xs =
	 List.fold_left (fun (a, b) -> fun h -> (a+h, b*h)) (0, 1) xs
	
let sumProd3 =
	List.fold_left ( fun acc num -> fst acc + num, snd acc * num ) (0, 1);;

(*zad4*)
(* zapętlanie się w wywolaniu rekursywnym quicksort large*)
(* szukamy wiekszych lub rowncyh elementów heada, z CAŁEJ	listy, łącznie z pierwszym elem*)
(* dlatego cała lista jest znów podawana jako argument (bo wszystkie są *)
(* większe lub równe), co tworzy nieskoncz. petle *)
 let rec quicksort = function
	 []  -> []          
	| [x] -> [x]          
	| xs  -> let small = List.filter (fun y -> y < List.hd xs ) xs
			and large = List.filter (fun y -> y >= List.hd xs ) xs
			in quicksort small @ quicksort large;;

(*usuwa duplikaty, ale sortuje dobrze*)
(* do large bierze elementy wieksze i mniejze od glowy, a rowne głowy ignoruje *)
let rec quicksort' = function
	[] -> []
	|  x::xs -> let small = List.filter (fun y -> y < x ) xs
		and large = List.filter (fun y -> y > x ) xs
		in quicksort' small @ (x :: quicksort' large);;

(*zad5*)
let isGreater a b = a > b

let rec insertionsort isGreater list =
	let rec insert element list =
		match list with
		| [] -> [element]
		| (h :: t) ->
				if isGreater element h then
					 	h :: (insert element t)
				else element :: h :: t
	in
	match list with
	| [] -> []
	| h :: t -> insert h (insertionsort isGreater t)

